### PR TITLE
Feature/exporter configuration

### PR DESCRIPTION
**Pull request by paulomach**
## Issue

Modifications to make exporter work properly within the snap.
Part of: [DPE-1255](https://warthogs.atlassian.net/browse/DPE-1255)

## Solution

1. Added dummy configure hook to enable config
2. Refactored exporter wrapper to embed default flags and get credentials

[DPE-1255]: https://warthogs.atlassian.net/browse/DPE-1255?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: snap/local/start-mysqld-exporter.sh [hunk 1]
```diff
@@ -1,6 +1,23 @@
...
-# For security measures, daemons should not be run as sudo. Execute mysqlrouter as the non-sudo user: snap-daemon.
-exec $SNAP/usr/bin/setpriv --clear-groups --reuid snap_daemon \
-  --regid snap_daemon -- $SNAP/bin/mysqld_exporter --config.my-cnf=$SNAP_DATA/etc/mysql/mysql.cnf "$@"
+USER="$(snapctl get exporter.user)"
+PASSWORD="$(snapctl get exporter.password)"
+SOCKET="$SNAP_COMMON/var/run/mysqld/mysqld.sock"
 
+# For security measures, daemons should not be run as sudo.
+# Execute mysqlrouter as the non-sudo user: snap-daemon.
+$SNAP/usr/bin/setpriv \
```
Comment (shayancanonical) [inline snap/local/start-mysqld-exporter.sh]: why not `exec`ing?

File: snap/snapcraft.yaml [hunk 2]
```diff
@@ -35,6 +35,8 @@ hooks:
     plugs:
       - network
       - network-bind
+  configure:
+    plugs: [network]
```
Comment (shayancanonical) [inline snap/snapcraft.yaml]: would prefer still a list with `-` to stay consistent with the above plugs

File: snap/snapcraft.yaml [hunk 3]
```diff
@@ -35,6 +35,8 @@ hooks:
     plugs:
       - network
       - network-bind
+  configure:
```
Comment (paulomach) [inline snap/snapcraft.yaml:55]: @jardon , even though the `configure` hook is a stub, I had to do add it here so `snap set <snap> key=value` would work.
I could not find in docs a better approach, but I'm open to modifications

File: snap/local/start-mysqld-exporter.sh (hunk 1 above)
Comment (paulomach) [reply snap/local/start-mysqld-exporter.sh]: ops, I must have accidentally removed it

File: snap/local/start-mysqld-exporter.sh [hunk 4]
```diff
@@ -1,6 +1,23 @@
 #!/bin/bash
 
-# For security measures, daemons should not be run as sudo. Execute mysqlrouter as the non-sudo user: snap-daemon.
-exec $SNAP/usr/bin/setpriv --clear-groups --reuid snap_daemon \
-  --regid snap_daemon -- $SNAP/bin/mysqld_exporter --config.my-cnf=$SNAP_DATA/etc/mysql/mysql.cnf "$@"
+USER="$(snapctl get exporter.user)"
```
Comment (taurus-forever) [inline snap/local/start-mysqld-exporter.sh]: LGTM. Q: Will this work in ROCK/OCI later?

File: snap/local/start-mysqld-exporter.sh (hunk 1 above)
Comment (paulomach) [reply snap/local/start-mysqld-exporter.sh]: 9674413

File: snap/snapcraft.yaml (hunk 2 above)
Comment (paulomach) [reply snap/snapcraft.yaml]: 82040d5

File: snap/local/start-mysqld-exporter.sh (hunk 4 above)
Comment (paulomach) [reply snap/local/start-mysqld-exporter.sh]: It won't.
We have two options:
1. Add more logic to the script, detecting if it's running from a SNAP, and case it's not, retrieve credentials, runtime user/group, from env's.
2. On the k8s  charm, configure the exporter service pebble layer to call `mysqld_exporter` binary directly, configuring parameters/env there.

tbh, given we are not just setting credentials, but also GUID, UID I prefer option 2.

What's your thoughts @taurus-forever ?

Comment (taurus-forever) [reply snap/local/start-mysqld-exporter.sh]: 2) is OK with me.

JFYI, [MongoDB](https://github.com/canonical/charmed-mongodb-snap/blob/5/edge/snap/local/start-mongodb-exporter.sh#L4). :-)

File: snap/snapcraft.yaml (hunk 3 above)
Comment (taurus-forever) [reply snap/snapcraft.yaml:55]: Ask snap team in MM?

**Review by taurus-forever — APPROVED** [review]

File: snap/local/start-mysqld-exporter.sh [hunk 5]
```diff
@@ -1,6 +1,23 @@
...
+PASSWORD="$(snapctl get exporter.password)"
+SOCKET="$SNAP_COMMON/var/run/mysqld/mysqld.sock"
```
Comment (jardon) [inline snap/local/start-mysqld-exporter.sh]: @paulomach  Does this have default behavior if the snap configs are not set?

Comment (paulomach) [reply snap/local/start-mysqld-exporter.sh]: not really - since this is the `charmed` snap i didn't mind. Should I?

Comment (paulomach) [reply snap/local/start-mysqld-exporter.sh]: its done

**Review by shayancanonical — APPROVED** [review]

**Pull request merged**